### PR TITLE
Stache tutorial (aka How to work with Mustache templates in Haskell)

### PR DESCRIPTION
**Pull request by mrkkrp**
Quite incomplete so far.

### CODE REVIEW

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+  instances). Source code is filled with Unicode symbols (using packages
+  like `base-unicode-symbols`, etc.), so it's not easy to edit it if you
+  want to contribute to the package.
+
+* [`stache`](https://hackage.haskell.org/package/stache) is an alternative
+  implementation that conforms to official Mustache spec and passes the
+  official test suite. Its API consists of only 4 functions (3 to compile
+  templates and one to render them using any instance of `ToJSON` as source
+  of data to interpolate), plus Template Haskell helpers to compile/validate
+  templates at compile time. Parser is written with Megaparsec 5. The
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "The parser is written with..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+  package uses `Data.Text.Lazy.Builder` under the hood and produces lazy
+  `Text`.
+
+The main motivation for developing `stache` was the desire to expose more
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "a more minimal..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+minimal API and use Aeson's instances directly for value interpolation, as
+well as the desire to use Megaparsec instead of Parsec for parsing.
+Initially we wanted to contribute to the existing `mustache` library, but
+then realized that changes we want to implement are too cardinal and we are
+better off writing our own package.
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "but then realized that the changes we wanted to implement were too cardinal and we were better off writing our own package."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+the sake of simplicity and better integration with Aeson.
+
+## Compiling templates
+
+Let's see how to compile templates because that's where you usually start.
+But first, let's take a look at types that will show up here and there in
+the rest of the tutorial:
+
+* `PName` stands for “partial's name”. “Partials” are other templates that
+  are inserted into actual template you're rendering. `PName` is a wrapper
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "the actual template you're rendering."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+    -- ^ Collection of all templates that are available for interpolation
+    -- (as partials). The top-level one is also contained here and the
+    -- “focus” can be switched easily by modifying 'templateActual'.
+  } deriving (Eq, Ord, Show, Data, Typeable, Generic)
+```
+
+When working with a `Template`, it's only possible to use partials  that are
+in `templateCache`, thus the main difference between compiling functions is
+where you get your template(s) and what you will have in `templateCache`.
+Thus there are three different ways to get a `Template`:
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "So, there are...", "therefore, there are...", or "thus, there are" (and "so the main difference..." above).

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+1. From lazy `Text` with `compileMustacheText`. The function takes a `PName`
+   and actual template source. The function returns `Either` parse error or
+   `Template` with cache consisting of only one `Template`.
+
+2. From a single `File` with `compileMustacheFile`. This one only takes path
+   to Mustache file the resulting template won't be able to use partials
+   unless you combine several templates into one (see below).
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "a path to a Mustache file" or "the path to a Mustache file." Also, there's something missing here (a period?) before "the resulting template won't be able to use..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+
+3. From a directory with `compileMustacheDir`. This function reads all
+   templates (files ending with `.mustache` extension) and puts them into
+   the cache, selecting one of them as main (you specify which). The
+   resulting `Template` can use partials that were present in that
+   directory.
+
+So to use partials, we need to use `compileMustacheDir` *or* combine several
+templates into one using `(<>)` method of `Semigroup` type class. Being an
+instance of the `Semigroup` type class means that you can always to two
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "combine two" or "append two" (or something other than "to"). Above ("_or_"), remember that the current font we have doesn't have an italic version.

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+import Data.Aeson
+import Data.Semigroup ((<>))
+import Data.Text (Text)
+import GHC.Generics
+import Text.Mustache
+import qualified Text.Mustache.Compile.TH as TH
+import qualified Data.Text.Lazy.IO as TIO
+```
+
+Then we will need to represent data somehow, for that let's define the
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "Then we will need to represent data somehow. For that, let's define..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+  } deriving (Eq, Show, Generic)
+
+instance ToJSON File
+```
+
+Here we have defined two records with various fields, just like we would do
+for other domains. We also have used
+[*generics*](https://wiki.haskell.org/GHC.Generics) to quickly define
+`ToJSON` instance for the data types. You can read more about generics, but
+the technique in a nutshell is simple: compiler can derive `Generic`
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "the compiler can derive..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+instance for your datatype (with `DeriveGeneric` extension enabled) which
+describes “structure” of your data, and then you can use that information to
+(again automatically) derive instance of `ToJSON` type class.
+
+Now we can experiment with template compilation. We will have two templates:
+`main.mustache` and `file.mustache` (we could do with just one, but then I
+wouldn't have a good example how to work with partials).
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "a good example of how to..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+REM DISCID "{{& reDiscId }}"
+REM COMMENT "{{& reComment }}"
+PERFORMER "{{& rePerformer }}"
+TITLE "{{& reTitle }}"
+{{# reFiles }}
+{{> file }}
+{{/ reFiles }}
+```
+
+Here we have layout of a typical header of a CUE file. The `{{> file }}`
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp "the typical layout..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+### Template Haskell helpers
+
+If you do not plan to change templates “on the fly” (without re-compiling
+your application), you may like Template Haskell (TH) helpers that compile
+Mustache templates at compile time and ensure that your templates are valid.
+If some template is not valid, your program just won't compile.
+
+The TH helpers live in the `Text.Mustache.Compile.TH` module and correspond
+precisely to the three normal functions we already know. Let's play with a
+template Haskell helper right in your editing environment.
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md:266]: @mrkkrp "right now"?

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+
+How does the magic work? What to do if you want to construct `Value`
+manually? Simply put, `toJSON` does not do anything you can't do yourself.
+Records are transformed into JSON `Object`s, lists and vectors and turned
+into `Array`s, etc. Feeding the `renderMustache` without `toJSON` is just as
+easy, consult
+[Aeson documentation](https://hackage.haskell.org/package/aeson) for more
+information, but the most common thing you will want to do is to create a
+custom dictionary that will be your “context”, it's done with help of
+`object` and `(.=)` functions:
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md:383]: @mrkkrp 
- "if you want to construct a `Value`..."
- "(consult Aeson's documentation for more information)" or "-- consult Aeson's documentation for more information --"
- "your context, which it's done with help of the..." or "your context. This is done with help of the..."

File: tutorials/haskell/mustache-templates/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+```
+
+It just couldn't be easier.
+
+## Conclusion
+
+`stache` seems to do its job just fine so far. We migrated some code from
+`hastache` and were surprised just how simpler the code looked with
+`stache`. It's also nice to be able to check your templates at compile time,
+like Shakespearean templates do. So in the conclusion I have to say that I
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md]: @mrkkrp I think "so in conclusion..." would be better here.

File: tutorials/haskell/mustache-templates/code/README.md [hunk 15]
```diff
@@ -0,0 +1,9 @@
+# Image processing
+
+Code for the “image processing” tutorial.
```
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/code/README.md]: @mrkkrp This one is not for image processing. :) Also, line 9 says BSD but the license file is MIT.

**Comment by jpvillaisaza** [comment]
@mrkkrp This looks very good. I left some comments for small changes.

File: tutorials/haskell/mustache-templates/tutorial.md (hunk 12 above)
Comment (mrkkrp) [inline tutorials/haskell/mustache-templates/tutorial.md:266]: @jpvillaisaza, “…right in your editing environment.” — I don't see what's wrong with this phrase.

**Comment by mrkkrp** [comment]
@jpvillaisaza, Your feedback is addressed.

File: tutorials/haskell/mustache-templates/tutorial.md (hunk 12 above)
Comment (jpvillaisaza) [inline tutorials/haskell/mustache-templates/tutorial.md:266]: @mrkkrp I don't know... I think it's good. I had a problem with the "right" when I first read it, but I can't think of any other option now.

**Comment by jpvillaisaza** [comment]
@mrkkrp Thanks! I think we can schedule this for Wednesday, August 17.

**Comment by jpvillaisaza** [comment]
@mrkkrp We can schedule this one for next Wednesday (August 17). Let's merge to tutorials so we can review on staging.

**Comment by mrkkrp** [comment]
@jpvillaisaza Sounds good, please go ahead.

**Comment by jpvillaisaza** [comment]
Done: https://staging.stackbuilders.com/tutorials/haskell/mustache-templates/

**Pull request merged**